src/stream/src/executor/source/fs_fetch_executor.rs

```diff
@@ -312,100 +336,85 @@ impl<S: StateStore, Src: OpendalSource> FsFetchExecutor<S, Src> {
-                                // of the NEXT message line in the file. However, In other source connectors,
-                                // we use the word `offset` to represent the offset of the current message.
-                                // We have to be careful about this semantical inconsistency.
-                                if offset.parse::<usize>().unwrap() >= fs_split.size {
-                                    splits_on_fetch -= 1;
-                                    state_store_handler.delete(split_id).await?;
-                                } else {
+                        // Motivation for the changes:
+                        //
+                        // Previously, the fetch executor determined whether a file was fully read by checking if the
+                        // offset reached the size of the file. However, this approach had some issues related to
+                        // maintaining the offset:
+                        //
+                        // 1. For files compressed with gzip, the size reported corresponds to the original uncompressed
+                        //    size, not the actual size of the compressed file.
+                        //
+                        // 2. For Parquet files, the offset represents the number of rows. Therefore, when listing each
+                        //    file, it was necessary to read the metadata to obtain the total number of rows in the file,
+                        //    which is an expensive operation.
+                        //
+                        // To address these issues, we changes the approach to determining whether a file is fully
+                        // read by moving the check outside the reader. The fetch executor's right stream has been
+                        // changed from a chunk stream to an `Option<Chunk>` stream. When a file is completely read,
+                        // a `None` value is added at the end of the file stream to signal to the fetch executor that the
+                        // file has been fully read. Upon encountering `None`, the file is deleted.
+                        Either::Right(optional_chunk) => match optional_chunk {
+                            Some(chunk) => {
+                                let mapping = get_split_offset_mapping_from_chunk(
+                                    &chunk, split_idx, offset_idx,
+                                )
+                                .unwrap();
+                                debug_assert_eq!(mapping.len(), 1);
+                                if let Some((split_id, _offset)) = mapping.into_iter().next() {
+                                    delete_file = split_id.clone();
+                                    let row = state_store_handler.get(split_id.clone()).await?
+                                        .unwrap_or_else(|| {
+                                            panic!("The fs_split (file_name) {:?} should be in the state table.",
```

Comment:
report error and skip the file? Seems the case is not that critical to the system, WDYT

Reply:
prefer to panic because it's not that we missed some files, but that some files that should have been read were deleted.